crates/matrix-sdk/src/widget/machine/from_widget.rs

```diff
@@ -41,28 +41,80 @@ pub(super) enum FromWidgetRequest {
     DelayedEventUpdate(UpdateDelayedEventRequest),
 }
 
+/// The full response a client sends to a from widget request
+/// in case of an error.
 #[derive(Serialize)]
 pub(super) struct FromWidgetErrorResponse {
     error: FromWidgetError,
 }
 
 impl FromWidgetErrorResponse {
-    pub(super) fn new(e: impl fmt::Display) -> Self {
-        Self { error: FromWidgetError { message: e.to_string() } }
+    /// Create a error response to send to the widget from an http error.
+    pub(crate) fn from_http_error(error: &HttpError) -> Self {
+        Self {
+            error: FromWidgetError {
+                message: error.to_string(),
+                matrix_api_error: error.as_client_api_error().and_then(
+                    |api_error| match &api_error.body {
+                        ErrorBody::Standard { kind, message } => Some(FromWidgetMatrixErrorBody {
+                            http_status: api_error.status_code.as_u16().into(),
+                            response: json!({"errcode": kind.to_string(), "error": message }),
+                        }),
+                        _ => None,
+                    },
+                ),
+            },
+        }
+    }
+
+    /// Create a error response to send to the widget from a matrix sdk error.
+    pub(crate) fn from_error(error: &Error) -> Self {
+        match &error {
+            Error::Http(e) => FromWidgetErrorResponse::from_http_error(e),
+            Error::UnknownError(e) => FromWidgetErrorResponse::from_string(e.to_string()),
+            _ => FromWidgetErrorResponse::from_string(error.to_string()),
+        }
+    }
+
+    /// Create a error response to send to the widget from a string.
+    pub(crate) fn from_string<S: Into<String>>(error: S) -> Self {
+        Self { error: FromWidgetError { message: error.into(), matrix_api_error: None } }
     }
 }
 
+/// The serializable section of an error response send by the client as a
+/// response to a from widget request.
 #[derive(Serialize)]
 struct FromWidgetError {
+    /// A unspecified error message text that caused this widget action to fail.
+    /// This is useful to prompt the user on an issue but cannot be used to
+    /// decide on how to deal with the error.
     message: String,
+    /// An optional matrix error that contains specified
+    /// information and helps finding a work around for specific errors.
+    matrix_api_error: Option<FromWidgetMatrixErrorBody>,
+}
+
+/// The serializable section of a widget response that represents a matrix
+/// error.
+#[derive(Serialize)]
+struct FromWidgetMatrixErrorBody {
+    /// The status code of the http response
+    http_status: u32,
+    /// The matrix standard error response including the `errorcode` and the
+    /// `error` message as defined in the spec: https://spec.matrix.org/v1.12/client-server-api/#standard-error-response
+    response: Value,
```

Comment:
Is it really worth saving the `response` field as a json::Value, instead of having the fields in there, automatically serialized by serde?

Reply:
I think using `StandardErrorBody` makes sense?
This struct was a bit suspicious to me see: https://matrix.to/#/!veagCdDBjKrMsOCzrq:privacytools.io/$2ieYTtk1E4HU36rQZ7r40_bjGaWUN-zfOGz01Yfv3Ng?via=matrix.org&via=envs.net&via=kabou.ro
(I was afraid it is only supposed to be used in a very special occasion since its not even used in the ErrorBody enum)
Also the `ErrorBody` enum seemed to be the perfect fit here but it is not serializable.
So I did not use it before.

Reply:
Would it make sense to store the result as an integer and the error message as a String?

Reply:
The errorcode is a a string as well. I used the StandardErrorBody which is exactly what we expect here.